libraries/chain/peer_keys_db.cpp

```diff
@@ -5,85 +5,46 @@ namespace eosio::chain {
 
 peer_keys_db_t::peer_keys_db_t() : _active(false) {}
 
-std::optional<public_key_type> peer_keys_db_t::get_peer_key(name n) const {
+peer_info_t peer_keys_db_t::get_peer_info(name n) const {
    fc::lock_guard g(_m);
-   if (auto it = _peer_key_map.find(n); it != _peer_key_map.end())
-      return std::optional<public_key_type>(it->second);
-   return std::optional<public_key_type>{};
+   assert(_active);
+   if (auto it = _peer_info_map.find(n); it != _peer_info_map.end())
+      return it->second;
+   return peer_info_t{};
 }
 
-size_t peer_keys_db_t::size() const {
-   fc::lock_guard g(_m);
-   return _peer_key_map.size();
-}
-
-// we update the keys that were registered up to lib_number (inclusive)
-// --------------------------------------------------------------------
-size_t peer_keys_db_t::update_peer_keys(const controller& chain, uint32_t lib_number) {
-   size_t num_updated = 0;
-   if (!_active || lib_number <= _block_num)
-      return num_updated;                      // nothing to do
-
-   try {
-      const auto& db = chain.db();
-      const auto  table_ref = boost::make_tuple("eosio"_n, "eosio"_n, "peerkeys"_n);
-      const auto* t_id      = db.find<table_id_object, by_code_scope_table>(table_ref);
-      EOS_ASSERT(t_id != nullptr, misc_exception, "cannot retrieve `peerkeys` table");
+peer_keys_db_t::new_peers_t peer_keys_db_t::update_peer_keys(const getpeerkeys_res_t& v) {
+   if (!_active || v.empty())
+      return {};
+   
+   // create hash_map of current top selected producers (according to "getpeerkeys"_n in system contracts)
+   // ----------------------------------------------------------------------------------------------------
+   peer_key_map_t current;
+   for (size_t i=0; i<v.size(); ++i)
+      current[v[i].producer_name] = peer_info_t{static_cast<uint32_t>(i), v[i].peer_key};
 
-      const auto& secidx = db.get_index<index64_index, by_secondary>();
-
-      const auto lower = secidx.lower_bound(std::make_tuple(t_id->id._id, static_cast<uint64_t>(_block_num + 1)));
-      const auto upper = secidx.upper_bound(std::make_tuple(t_id->id._id, static_cast<uint64_t>(lib_number)));
-
-      if (upper == lower) {
-         // no new keys registered
-         _block_num = lib_number;
-         return num_updated;
+   fc::lock_guard g(_m);
+   new_peers_t res;
+
+   // update ranking of those removed from top selected producers
+   // -----------------------------------------------------------
+   for (auto& pi : _peer_info_map) {
+      if (!current.contains(pi.first)) {
+         pi.second.rank = std::numeric_limits<uint32_t>::max();
+         res.insert(pi.first);
```

Comment:
I think you should remove the non-top ones from `_peer_info_map`. This can return ones that are no longer in the list as modified, but I don't think we need to keep them around.

Reply:
https://github.com/AntelopeIO/spring/pull/1316/commits/de28545b8fe0a231beba73a7fa2e9e2fc8315568